README.md
# TCPDup
...a duplicator for TCP streams.

## Why would you need this?
In a clustered Icinga2 setup, when exporting metrics using the Graphite module,
each performance metric is only written by a single master. If you now want to
duplicate the data to all masters, you'll need to duplicate the stream, which
is why we have TCPDup.
TCPDup simply listens on a socket, accepts connections and will send all data
to all backends, responses from a backend are silently ignored.

## Usage
```
You need to specify at least one backend!
Options:
  --help                  print this message
  --backend arg           backend to use ("x.x.x.x:port")
  --buf arg (=8192)       Number of measurements to buffer in case of
                          connectivity issues
  --port arg (=12345)     port to bind to
  --bind arg (=127.0.0.1) IP to bind to
```
Backend should probably be specified multiple times, the rest is pretty
self-explanatory

## Building
Assuming that you don't want to copy the binary around, you can build a debian
package using
```
dpkg-buildpackage -us -uc
```
The command-line options are then set in `/etc/default/tcpdup` and a systemd
unit is included.


debian/changelog
tcpdup (0.1.1) unstable; urgency=medium

  * Adjust seccomp filter for newer stdlibs

 -- Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>  Mon, 22 Jul 2019 18:46:24 +0200

tcpdup (0.1+nmu1) unstable; urgency=medium

  * Experimental seccomp support
  * Fix unit file to limit memory
  * Print buffer fill status every 10 messages

 -- Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>  Sun, 22 Oct 2017 14:21:37 +0200

tcpdup (0.1) unstable; urgency=medium

  * Initial Release

 -- Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>  Wed, 12 Apr 2017 09:00:01 +0200


TCPClient.h
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#ifndef TCPDUP_TCPCLIENT_H
#define TCPDUP_TCPCLIENT_H

#include <boost/asio.hpp>
#include <deque>
#include <string>

class TCPClient {
public:
    TCPClient(boost::asio::io_service& io_service, boost::asio::ip::tcp::resolver::iterator& endpoint_iterator,
              int backendNumber, unsigned int buf_size);
    void sendMessage(std::string & message);

private:
    bool connected;
    boost::asio::ip::tcp::resolver::iterator endpoint_iterator;
    boost::asio::io_service& io_service;
    boost::asio::deadline_timer timer;
    std::unique_ptr<boost::asio::ip::tcp::socket> socket;
    std::deque<std::string> messages, offline_messages;
    int backend_number;
    unsigned int buf_size;

    void do_connect();
    void do_write();
    void reconnect_later();
};


#endif //TCPDUP_TCPCLIENT_H


TCPServer.h
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#ifndef TCPDUP_TCPSERVER_H
#define TCPDUP_TCPSERVER_H

#include <boost/asio.hpp>
#include "TCPServerSession.h"

class TCPServer {
public:
    TCPServer(boost::asio::io_service& io_service, const boost::asio::ip::tcp::endpoint& endpoint,
              std::list<std::shared_ptr<TCPClient>>& clients);

private:
    void start_accept();
    void handle_accept(std::shared_ptr<TCPServerSession> session, const boost::system::error_code &ec);

private:
    boost::asio::io_service& io_service;
    boost::asio::ip::tcp::acceptor acceptor;
    std::list<std::shared_ptr<TCPClient>> clients;
};


#endif //TCPDUP_TCPSERVER_H


TCPClient.cpp
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#include <boost/asio.hpp>
#include <boost/bind.hpp>
#include <deque>
#include <iostream>
#include <memory>
#include "TCPClient.h"

using boost::asio::ip::tcp;

TCPClient::TCPClient(boost::asio::io_service& io_service, tcp::resolver::iterator& endpoint_iterator,
                     int backendNumber, unsigned int buf_size):
            io_service (io_service), timer(io_service), backend_number(backendNumber), buf_size(buf_size) {
    this->endpoint_iterator = std::move(endpoint_iterator);
    do_connect();
}

void TCPClient::sendMessage(std::string & message) {
    if (!connected) {
        if (offline_messages.size() < buf_size) {
            offline_messages.push_back(message);
        }
        if (offline_messages.size() % 10 == 0) {
            std::cerr << "Backend " << backend_number << ", " << offline_messages.size()  << " of " << buf_size
                      << " (max) messages in offline buffer" << std::endl;
        }
    } else {
        io_service.post(
                [this, message]() {
                    // This is run on the io_service and therefore cannot race!
                    bool stillWriting = !messages.empty();
                    messages.push_back(message);
                    if (!stillWriting) {
                        do_write();
                    }
                });
    }
}

void TCPClient::do_connect() {
    // re-create socket for new connection
    socket = std::move(std::make_unique<tcp::socket>(io_service));

    boost::asio::async_connect(*socket, endpoint_iterator,
                               [this](boost::system::error_code ec, tcp::resolver::iterator) {
                                   if (!ec) {
                                       connected = true;
                                       std::cerr << "Backend " << backend_number << ", connected" << std::endl;
                                       io_service.post(
                                               [this]() {
                                                   // This is run on the io_service and therefore cannot race!
                                                   while (!offline_messages.empty()) {
                                                       messages.push_back(offline_messages.front());
                                                       offline_messages.pop_front();
                                                   }
                                                   if (!messages.empty()) {
                                                       do_write();
                                                   }
                                               });
                                   } else {
                                       std::cerr << "Backend " << backend_number << ", coudln't connect to backend: "
                                                 << ec.message() << std::endl;
                                       socket->close();
                                       reconnect_later();
                                   }
                               });
}

void TCPClient::do_write() {
    if (!connected)
        return;
    boost::asio::async_write(*socket,
                             boost::asio::buffer(messages.front().data(),
                                                 messages.front().length()),
                             [this](boost::system::error_code ec, std::size_t) {
                                 if (!ec && connected) {
                                     // Remove successfully sent message from queue and
                                     // recursively call this function to send the next one
                                     messages.pop_front();
                                     if (!messages.empty()) {
                                         do_write();
                                     }
                                 }
                                 else {
                                     std::cerr << "Backend " << backend_number << ", error during write: "
                                               << ec.message() << std::endl;
                                     connected = false;
                                     socket->close();
                                     reconnect_later();
                                 }
                             });
}

void TCPClient::reconnect_later() {
    if (!connected) {
        // Reconnect after 5 second
        std::cerr << "Backend " << backend_number << ", reconnecting in 5 seconds" << std::endl;
        timer.expires_from_now(boost::posix_time::seconds(5));
        timer.async_wait(std::bind(&TCPClient::do_connect, this));
    }
}


TCPServer.cpp
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#include <boost/asio.hpp>
#include <boost/bind.hpp>
#include <iostream>
#include "TCPServer.h"

using boost::asio::ip::tcp;

TCPServer::TCPServer(boost::asio::io_service& io_service, const tcp::endpoint& endpoint,
              std::list<std::shared_ptr<TCPClient>>& clients) :
            io_service(io_service), acceptor(io_service, endpoint), clients(clients) {
    start_accept();
}

void TCPServer::start_accept() {
    std::shared_ptr<TCPServerSession> new_session(new TCPServerSession(io_service, clients));
    acceptor.async_accept(new_session->socket(),
                          boost::bind(&TCPServer::handle_accept, this, new_session,
                                      boost::asio::placeholders::error));
}

void TCPServer::handle_accept(std::shared_ptr<TCPServerSession> session,
                              const boost::system::error_code &ec) {
    if (!ec) {
        session->start();
    } else {
        std::cerr << "Error while accepting connection: " << ec.message() << std::endl;
    }

    start_accept();
}


TCPServerSession.h
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#ifndef TCPDUP_TCPSERVERSESSION_H
#define TCPDUP_TCPSERVERSESSION_H

#include <boost/asio.hpp>
#include <list>
#include "TCPClient.h"

class TCPServerSession : public std::enable_shared_from_this<TCPServerSession>
{
public:
    TCPServerSession(boost::asio::io_service& io_service, std::list<std::shared_ptr<TCPClient>>& clients);
    boost::asio::ip::tcp::socket& socket();
    void start();
    void do_read(const boost::system::error_code &error);

private:
    boost::asio::ip::tcp::socket my_socket;
    boost::asio::streambuf buffer;
    std::list<std::shared_ptr<TCPClient>> clients;
};

#endif //TCPDUP_TCPSERVERSESSION_H


TCPServerSession.cpp
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#include <boost/asio.hpp>
#include <boost/bind.hpp>
#include <deque>
#include <iostream>
#include "TCPServerSession.h"
#include "TCPClient.h"

using boost::asio::ip::tcp;

TCPServerSession::TCPServerSession(boost::asio::io_service& io_service,
                     std::list<std::shared_ptr<TCPClient>>& clients)
            : my_socket(io_service), clients(clients) {
}

boost::asio::ip::tcp::socket& TCPServerSession::socket() {
    return my_socket;
}

void TCPServerSession::start() {
    boost::asio::async_read_until(my_socket,
                                  buffer,
                                  '\n',
                                  boost::bind(&TCPServerSession::do_read, shared_from_this(),
                                              boost::asio::placeholders::error));
}

void TCPServerSession::do_read(const boost::system::error_code &error) {
    if (!error) {
        std::istream is(&buffer);
        std::string line;
        std::getline(is, line);
        line += "\n";
	for (const auto client : clients) {
            client->sendMessage(line);
        }
        boost::asio::async_read_until(my_socket,
                                      buffer,
                                      '\n',
                                      boost::bind(&TCPServerSession::do_read, shared_from_this(),
                                                  boost::asio::placeholders::error));
    }
    else {
        std::cerr << "Error during read from client, closing connection!" << std::endl;
        my_socket.close();
    }
}


main.cpp
/*
 * Copyright (C) 2017  Maximilian Falkenstein <mfalkenstein@sos.ethz.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
#include <iostream>
#include <boost/asio.hpp>
#include <boost/program_options.hpp>
#include <boost/regex.hpp>

#include <sys/prctl.h>
#include <seccomp.h>

#include "TCPClient.h"
#include "TCPServer.h"

using boost::asio::ip::tcp;
namespace po = boost::program_options;

int main(int argc, const char* const* argv) {
    // https://stackoverflow.com/a/25969006
    boost::regex ip_regex("^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$");

    // Not to be confused with option_description which is something completely different
    po::options_description desc("Options");
    desc.add_options()
            ("help", "print this message")
            ("backend", po::value<std::vector<std::string>>()->multitoken(),
             "backend to use (\"x.x.x.x:port\")")
            ("buf", po::value<unsigned int>()->default_value(8192), "Number of measurements to buffer in case of connectivity issues")
            ("port", po::value<unsigned short>()->default_value(12345), "port to bind to")
            ("bind", po::value<std::string>()->default_value("127.0.0.1"), "IP to bind to")
            ("seccomp-dbg", "Trap instead of exit on seccomp violation (debug only!)")
            ;
    po::variables_map options;
    po::store(po::parse_command_line(argc, argv, desc), options);
    po::notify(options);

    // Check options
    if (options.count("help")) {
        std::cout << desc << std::endl;
        return 0;
    }

    // Check if at least one backend was specified
    if (!options.count("backend")) {
        std::cerr << "You need to specify at least one backend!" << std::endl;
        std::cout << desc << std::endl;
        return -1;
    }

    // Check if the bind address makes sense
    std::string bindAddr = options["bind"].as<std::string>();
    if (!boost::regex_match(bindAddr, ip_regex)) {
        std::cerr << "Bind address has invalid format: " << bindAddr << std::endl;
        return -1;
    }

    // Init seccomp filter
    scmp_filter_ctx ctx;
    if (options.count("seccomp-dbg")) {
        ctx = seccomp_init(SCMP_ACT_TRAP); // default action: trap
    } else {
        ctx = seccomp_init(SCMP_ACT_KILL); // default action: kill
        prctl(PR_SET_NO_NEW_PRIVS, 1);     // disallow setting permissive seccomp rules afterwars
        prctl(PR_SET_DUMPABLE, 0);         // disallow changing settings via ptrace from another process
    }

    // Basics
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(exit), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(read), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(write), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(munmap), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(brk), 0);
    //seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mprotect), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(exit_group), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(epoll_wait), 0);
    // boost::asio
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(poll), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(getsockopt), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(accept), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(epoll_ctl), 0);
    // ioctl is needed to enable async io on a newly established connection
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(ioctl), 1, SCMP_A1(SCMP_CMP_EQ, FIONBIO));
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(recvmsg), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(sendmsg), 0);
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(close), 0);
    // timers
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(timerfd_settime), 0);
    // reconnect
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(socket), 0); // TODO: restrict if possible
    seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(connect), 0); // TODO: restrict if possible

    // Init ASIO
    boost::asio::io_service ioservice;
    tcp::resolver resolver(ioservice);

    // Open TCP server socket
    tcp::endpoint tcp_endpoint{boost::asio::ip::address::from_string(options["bind"].as<std::string>()),
                                options["port"].as<unsigned short>()};

    // Open TCP client connection to each backend
    std::vector<std::string> backends = options["backend"].as<std::vector<std::string>>();
    auto clientList = std::list<std::shared_ptr<TCPClient>>();
    std::cout << "Number of backends: " << backends.size() << std::endl;

    for (int i = 0; i < backends.size(); i++) {
        std::string backend = backends[i];

        auto index = backend.find(':');
        if (index > 0 && index<backend.size()-1) {
            auto host = backend.substr(0, index);
            auto port = backend.substr(index+1);
            std::cout << "Backend " << i << ": Host: " << host << ", Port: " << port << std::endl;

            tcp::resolver::query query(tcp::v4(), host, port);
            tcp::resolver::iterator iterator = resolver.resolve(query);
            clientList.push_back(std::make_shared<TCPClient>(ioservice, iterator, i, options["buf"].as<unsigned int>()));
        } else {
            std::cout << "Backend " << i << " has invalid format (" << backend << "), ignored!" << std::endl;
        }
    }

    TCPServer server(ioservice, tcp_endpoint, clientList);
    /*
     * Boost ASIO's timers will open /etc/localtime (once!). open() is very difficult to restrict using seccomp's BPF,
     * therefore we simply execute a timer before enabling the seccomp rules.
     * */
    boost::asio::deadline_timer timer(ioservice);
    timer.expires_from_now(boost::posix_time::seconds(1));
    timer.async_wait([ctx] (const boost::system::error_code&) {
        seccomp_load(ctx);
    });
    ioservice.run();
}
